ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala

```diff
@@ -0,0 +1,134 @@
+  private[http] def transactionMessageHandler(jwt: Jwt, jwtPayload: JwtPayload): Flow[Message, Message, _] = {
+    import scala.concurrent.duration._
+
+    wsMessageHandler(jwt, jwtPayload)
+      .takeWithin(wsConfig.maxDuration)
+      .throttle(wsConfig.throttleElem,
+                wsConfig.throttlePer,
+                wsConfig.maxBurst,
+                wsConfig.mode)
+      .keepAlive(5.seconds, () => TextMessage.Strict(heartBeat))
+      .watchTermination() {
+        (_, future) =>
+          numConns.incrementAndGet
+          logger.info(s"New websocket client has connected, current number of clients:$numConns")
+          future onComplete {
+            case Success(_) =>
+              numConns.decrementAndGet
+              logger.info(s"Websocket client has disconnected. Current number of clients: $numConns")
+            case Failure(ex) =>
+              numConns.decrementAndGet
+              logger.info(s"Websocket client interrupted on Failure: ${ex.getMessage}. remaining number of clients: $numConns")
+          }
+      }
+  }
+
+  private def wsMessageHandler(jwt: Jwt, jwtPayload: JwtPayload): Flow[Message, Message, NotUsed] = {
+    Flow[Message]
+      .flatMapConcat {
+        case msg: TextMessage.Strict => generateOutgoingMessage(jwt, jwtPayload, msg)
+        case _ => Source.single(wsErrorMessage("Cannot process your input, Expect a single strict JSON message"))
+      }
+  }
+
+  private def generateOutgoingMessage(jwt: Jwt, jwtPayload: JwtPayload, incoming: TextMessage.Strict): Source[Message, NotUsed] = {
+    val maybeIncomingJs = SprayJson.parse(incoming.text).toOption
+    parseActiveContractsRequest(maybeIncomingJs)
+      .leftMap(e => InvalidUserInput(e.shows)) match {
+      case \/-(req) => getTransactionSourceForParty(jwt, jwtPayload, req)
+      case -\/(e) => Source.single(wsErrorMessage(s"Error happend parsing your input message to a valid Json request: $e"))
+    }
+  }
+
+  private def parseActiveContractsRequest(
+                                           incoming: Option[JsValue]
+                                         ): SprayJson.JsonReaderError \/ GetActiveContractsRequest = {
+    incoming match {
+      case Some(JsObject.empty) => \/-(emptyGetActiveContractsRequest)
+      case Some(jsObj) => SprayJson.decode[GetActiveContractsRequest](jsObj)
+      case None => -\/(JsonReaderError("None", "please send a valid json request"))
+    }
+  }
+
+  private def getTransactionSourceForParty(jwt: Jwt, jwtPayload: JwtPayload, request: GetActiveContractsRequest): Source[Message, NotUsed] = {
+    import com.digitalasset.http.util.Transactions._
+
+    resolveTemplateIds(request.templateIds) match {
+      case \/-(ids) =>
+        val filter = transactionFilterFor(jwtPayload.party, ids)
+        transactionClient.getTransactions(LedgerOffsetOrdering.ledgerBegin, None, transactionFilter = filter) // TODO: make offSet pass along with client message
```

Comment:
we probably don't want to consume all transactions from the very beginning all the time.

Reply:
yeah, same thoughts on this too, (see TODO the comment), maybe we can wrap the client request message with offset input